Migrate custom-flows 'Overview' page from Sanity to MDX (#297)
clerkinc/clerk-docs: docs/custom-flows/overview.mdx

---
title: Custom flows
description: Learn the process behind building custom sign-up and sign-in flows with Clerk.
---

# Custom flows

> If Clerk's [prebuilt components](/docs/components/overview) don't meet your specific needs or if you require more control over the authentication flow, Clerk enables you to build fully custom sign-up and sign-in flows using our [`useSignUp()`](/docs/references/react/use-sign-up) and [`useSignIn()`](/docs/references/react/use-sign-in) React hooks.

Reviewer: ```suggestion
If Clerk's [prebuilt components](/docs/components/overview) don't meet your specific needs or if you require more control over the authentication flow, Clerk enables you to build fully custom sign-up and sign-in flows using its [`useSignUp()`](/docs/references/react/use-sign-up) and [`useSignIn()`](/docs/references/react/use-sign-in) React hooks.
```

See my Slack message about not using "we"/"our"

> But before diving deep into the code side, it is important to first understand the process behind building sign-up and sign-in flows, which fields are required, and the necessary steps to authenticate a user.

Reviewer: ```suggestion
For your general understanding it's important to know the process behind building sign-up and sign-in flows. In this document you'll learn which fields are required and which steps are necessary to authenticate a user. Afterwards you're ready to dive into the other related guides.
```

Shorter sentences make it clearer I think

## Sign-up flow

> Clerk provides a flexible way to build sign-up flows in your application. You can use a single [`SignUp`](/docs/references/javascript/sign-up/sign-up) object to gather information, verify their email address or phone number, add OAuth accounts, and finally, convert them into a [`User`](/docs/references/javascript/user/user).

Reviewer: Suggestion: Since the links repeat a lot you could use https://www.markdownguide.org/basic-syntax/#reference-style-links

This way you'd e.g. write

```md
[`SignUp`][sign-up-ref]

blabla...

[sign-up-ref]: /docs/references/javascript/sign-up/sign-up
```

Every [`SignUp`](/docs/references/javascript/sign-up/sign-up) has a set of requirements it must meet before it is turned into a [`User`](/docs/references/javascript/user/user). These requirements are defined by the instance settings you selected in the [Clerk Dashboard](https://dashboard.clerk.com/). Once all requirements are met, the [`SignUp`](/docs/references/javascript/sign-up/sign-up) will turn into a new [`User`](/docs/references/javascript/user/user), and an active session for that [`User`](/docs/references/javascript/user/user) will be created on the current [`Client`](/docs/references/javascript/client).

Don't worry about collecting all the required fields at once and passing them to a single request. The API is designed to accommodate a progressive sign-up flow, often corresponding to multi-step sign-up forms.

### Required fields

> The [`SignUp`](/docs/references/javascript/sign-up/sign-up) will show the current state of requirement collection. You can consult the `required_fields`, `optional_fields`, and `missing_fields` keys for a hint on where things are and what you need to do next.

Reviewer: The first sentence here is a bit confusing to me. What is a "requirement collection"?

| `verifications` | An object that describes the current state of verification for the [`SignUp`](/docs/references/javascript/sign-up/sign-up). There are currently three different keys: `email_address`, `phone_number`, and `external_account`. |

## Sign-in flow

Sign-in's are initiated by creating a [`SignIn`](/docs/references/javascript/sign-in/sign-in) object on the current [`Client`](/docs/references/javascript/client). The [`SignIn`](/docs/references/javascript/sign-in/sign-in) handles all the state and logic associated with a sign-in. If the sign-in is successfully authenticated, it will transform into an active session on the current [`Client`](/docs/references/javascript/client).

### Completing a sign-in

There are 3 main steps a user must perform in order to complete a sign-in.

<Steps>

### Identification

The first step a user needs to make is to identify what account they'd like to sign in to. This is done with an identifier, which can either be an email address, a phone number, or a username.

### Factor one verification

> Once a user is identified, they need to prove that "they are who they say they are". This is the process of "authenticating" the user. There's a number of strategies a user can use to perform authentication with the most basic being the humble password. Other authentication strategies, like passwordless sign-in, can be explored in the [sign-in options documentation](/docs/authentication/configuration/sign-in-options). 

Reviewer: ```suggestion
Once a user is identified, they need to prove their identity. This is the process of authenticating the user. There's a number of strategies a user can use to perform authentication with the most basic being the password. Other authentication strategies, like passwordless sign-in, can be explored in the [sign-in options documentation](/docs/authentication/configuration/sign-in-options). 
```

The paragraph was a bit too colloquial for my taste
